mbedtls/src/ssl/context.rs

```diff
@@ -251,62 +241,152 @@ impl<'a> Session<'a> {
         }
     }
 
+
+    // Session specific functions
+    
     /// Return the 16-bit ciphersuite identifier.
     /// All assigned ciphersuites are listed by the IANA in
     /// https://www.iana.org/assignments/tls-parameters/tls-parameters.txt
-    pub fn ciphersuite(&self) -> u16 {
+    pub fn ciphersuite(&self) -> Result<u16> {
         if self.inner.session == ::core::ptr::null_mut() {
-            0
-        } else {
-            unsafe { self.inner.session.as_ref().unwrap().ciphersuite as u16 }
+            return Err(Error::SslBadInputData);
         }
+        
+        Ok(unsafe { self.inner.session.as_ref().unwrap().ciphersuite as u16 })
     }
 
-    pub fn peer_cert(&self) -> Option<crate::x509::certificate::Iter> {
-        unsafe { UnsafeFrom::from(ssl_get_peer_cert(self.inner)) }
+    // This makes a copy of the certificate as peer_cert may be destroyed outside our control.
+    pub fn peer_cert(&self) -> Result<Option<&MbedtlsList<Certificate>>> {
+        if self.inner.session == ::core::ptr::null_mut() {
+            return Err(Error::SslBadInputData);
+        }
+
+        unsafe {
+            let peer_cert : &MbedtlsList<Certificate> = UnsafeFrom::from(&((*self.inner.session).peer_cert) as *const *mut x509_crt as *const *const x509_crt).ok_or(Error::SslBadInputData)?;
+            Ok(Some(peer_cert))
+        }
     }
+}
 
-    pub fn verify_result(&self) -> StdResult<(), VerifyError> {
-        match unsafe { ssl_get_verify_result(self.inner) } {
-            0 => Ok(()),
-            flags => Err(VerifyError::from_bits_truncate(flags)),
+impl Drop for Context {
+    fn drop(&mut self) {
+        unsafe {
+            ssl_close_notify(self.into());
```

Comment:
FYI: I'm guessing this may do IO? This may need revisiting once we implement async IO.

Reply:
correct, it can do IO.